### PR TITLE
frexp improvements

### PR DESCRIPTION
**Pull request by musm**
the bit shift left should  an unsigned int

### CODE REVIEW

File: base/math.jl [hunk 1]
```diff
@@ -413,19 +413,17 @@ and `val` is equal to ``x \\times 2^{exp}``.
 function frexp{T<:AbstractFloat}(x::T)
     xu = reinterpret(Unsigned,x)
     xe = xu & exponent_mask(T)
+    (xe == exponent_mask(T) || x == 0) && return x, 0 # +-0, NaN, or Inf
     k = Int(xe >> significand_bits(T))
     if xe == 0 # x is subnormal
-        x == 0 && return x, 0
```
Comment (musm) [inline base/math.jl]: no reason to check x == 0 here within the xe==0. Moved these up top to reduce redundancy with the xe == 0 check.

**Comment by musm** [comment]
Alternatively one could also scale the subnormal case through a multiplication as done in the openlibm implementation.

File: base/math.jl [hunk 2]
```diff
@@ -411,23 +411,23 @@ Return `(x,exp)` such that `x` has a magnitude in the interval ``[1/2, 1)`` or 0
...
         xs = xu & sign_mask(T)
         xu $= xs
-        m = leading_zeros(xu)-exponent_bits(T)
+        m = (leading_zeros(xu) - exponent_bits(T)) % UInt
         xu <<= m
         xu $= xs
-        k = 1-m
-    elseif xe == exponent_mask(T) # NaN or Inf
-        return x,0
+        k = 1 - m % Int
```
Comment (simonbyrne) [inline base/math.jl]: Maybe add parentheses: `1 - (m % Int)`

**Comment by musm** [comment]
Done,  is there any chance we will see a different way to do  `% Int` . IMO it's  really bad to have to do integer truncation like this

File: base/math.jl [hunk 3]
```diff
@@ -411,25 +411,25 @@ Return `(x,exp)` such that `x` has a magnitude in the interval ``[1/2, 1)`` or 0
...
+    xu = reinterpret(Unsigned, x)
     xe = xu & exponent_mask(T)
+    xe == exponent_mask(T) && return x, 0 # NaN or Inf
     k = Int(xe >> significand_bits(T))
     if xe == 0 # x is subnormal
-        x == 0 && return x, 0
         xs = xu & sign_mask(T)
         xu $= xs
-        m = leading_zeros(xu)-exponent_bits(T)
+        xu == 0 && return x, 0 # +-0
```
Comment (musm) [inline base/math.jl]: Since we already have the  bit representation, it's better to   check the zero here as opposed to `x == 0`

**Comment by stevengj** [comment]
To be clear, these are just performance optimizations, right?  Do you have any benchmark of the impact of your changes?

**Comment by musm** [comment]
Just perf optims, that I can across thought other work, when checking the frexp code.

Well previously  `m` was a signed int even though  `m >0`, so the shift is is slow.
Its also faster to compare `xe` as an int to the zero since we already masked out the sign bit than the float  ` x == 0`
For the exceptional case `xe == exponent_mask(T) # NaN or Inf` best to check at top,  instead of computing `k` first since we don't `k` in this case

**Comment by stevengj** [comment]
How much faster is it?

**Comment by musm** [comment]
For subnormal case the median is about 2-3 ns faster, for other cases it's basically the same. The total time such operations is (between  4-5 ns).

**Comment by stevengj** [comment]
@musm, please do actual benchmarks, don't just estimate according to what you think the cost of the operations is theoretically.  Theoretical performance estimates are notoriously unreliable on real computers.

I just benchmarked your implementation, and it seems to be type-unstable (`@code_warntype` says that the return type is `Tuple{Any,Any}`) and hence horrendously slow.

**Comment by musm** [comment]
I did.

Not sure how you are getting type instabilities. Try, 
`
using Base: exponent_half, exponent_mask, significand_bits, leading_zeros, exponent_bits, sign_mask,exponent_bias 
`
before testing.

**Comment by simonbyrne** [comment]
> Done, is there any chance we will see a different way to do % Int . IMO it's really bad to have to do integer truncation like this

In this case you could also use `unsigned` and `signed`.

**Comment by stevengj** [comment]
I already imported those (the code wouldn't run without it):
```jl
import Base.Math: exponent_mask, significand_bits, exponent_bias, exponent_half
function myfrexp{T<:AbstractFloat}(x::T) # copy-and-pasted from the PR
           xu = reinterpret(Unsigned, x)
           xe = xu & exponent_mask(T)
           xe == exponent_mask(T) && return x, 0 # NaN or Inf
           k = Int(xe >> significand_bits(T))
           if xe == 0 # x is subnormal
               xs = xu & sign_mask(T)
               xu $= xs
               xu == 0 && return x, 0 # +-0
               m = (leading_zeros(xu) - exponent_bits(T)) % UInt
               xu <<= m
               xu $= xs
               k = 1 - (m % Int)
           end
           k -= (exponent_bias(T) - 1)
           xu = (xu & ~exponent_mask(T)) | exponent_half(T)
           return reinterpret(T, xu), k
end
@code_warntype myfrexp(1.34)
```
gives (in Julia 0.5):
```
Variables:
  #self#::#myfrexp
  x::Float64
  xe::Any
  xs::Any
  m::Any
  k::Any
  xu::Any

Body:
  begin 
      NewvarNode(:(xs::Any))
      NewvarNode(:(m::Any))
      NewvarNode(:(k::Any))
      xu::Any = (Base.box)(UInt64,x::Float64) # line 3:
      xe::Any = (xu::UInt64 & (Main.exponent_mask)($(Expr(:static_parameter, 1)))::Any)::Any # line 4:
      unless (xe::Any == (Main.exponent_mask)($(Expr(:static_parameter, 1)))::Any)::Any goto 10
      return (Core.tuple)(x::Float64,0)::Tuple{Float64,Int64}
      10:  # line 5:
      SSAValue(0) = (xe::Any >> (Main.significand_bits)($(Expr(:static_parameter, 1)))::Any)::Any
      k::Any = (Core.typeassert)((Base.convert)(Int64,SSAValue(0))::Any,Int64)::Int64 # line 6:
      unless (xe::Any == 0)::Any goto 32 # line 7:
      xs::Any = (xu::UInt64 & (Main.sign_mask)($(Expr(:static_parameter, 1)))::Any)::Any # line 8:
      xu::Any = (xu::UInt64 $ xs::Any)::Any # line 9:
      unless (xu::Any == 0)::Any goto 23
      return (Core.tuple)(x::Float64,0)::Tuple{Float64,Int64}
      23:  # line 10:
      m::Any = (((Main.leading_zeros)(xu::Any)::Int64 - (Main.exponent_bits)($(Expr(:static_parameter, 1)))::Any)::Any % Main.UInt)::Any # line 11:
      xu::Any = (xu::Any << m::Any)::Any # line 12:
      xu::Any = (xu::Any $ xs::Any)::Any # line 13:
      k::Any = (1 - (m::Any % Main.Int)::Any)::Any
      32:  # line 15:
      k::Any = (k::Any - ((Main.exponent_bias)($(Expr(:static_parameter, 1)))::Any - 1)::Any)::Any # line 16:
      xu::Any = ((xu::Any & ~((Main.exponent_mask)($(Expr(:static_parameter, 1)))::Any)::Any)::Any | (Main.exponent_half)($(Expr(:static_parameter, 1)))::Any)::Any # line 17:
      return (Core.tuple)((Main.reinterpret)($(Expr(:static_parameter, 1)),xu::Any)::Any,k::Any)::Tuple{Any,Any}
  end::Tuple{Any,Any}
```
Does it only work with Julia 0.6?

**Comment by stevengj** [comment]
Okay, now it is type-stable. It seems to be about 8% slower for `x = rand(10^7)` (the non-subnormal case), and about 9x faster(!!) for the subnormal case `x = rand(10^7)/1e-308`.

I did a simple benchmark of
```jl
foo(x) = mapreduce(x -> frexp(x)[1], +, 0.0, x)
using BenchmarkTools
@benchmark foo($x)
```

**Comment by musm** [comment]
g2g

**Comment by musm** [comment]
bump

**Comment by stevengj** [comment]
Do we have good test coverage of this function, covering all the branches?

**Comment by musm** [comment]
I'm inclined to close this, since I'd rather write the tests and functions in my own package and carry development there instead.

**Comment by stevengj** [comment]
Since this function is in Base, it would be good to fix the one in Base, and it seems like you have a good improvement here.  But I just wanted to make sure we have test coverage of changed code before it can be merged.

I looked in `test/math.jl`, and it seems that `frexp` is indeed being tested in the subnormal case, so I think we are good to go.

**Pull request merged**